File: src/main/java/com/nanabell/sponge/nico/discordlink/event/LinkEventContextKeys.java
package com.nanabell.sponge.nico.discordlink.event;

import com.nanabell.sponge.nico.discordlink.LinkResult;
import net.dv8tion.jda.api.entities.MessageChannel;
import net.dv8tion.jda.api.entities.User;
import org.spongepowered.api.event.cause.EventContextKey;
import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;

public class LinkEventContextKeys {

    public static final EventContextKey<User> USER = createFor("DISCORD_USER");
    public static final EventContextKey<MessageChannel> MESSAGE_CHANNEL = createFor("DISCORD_MESSAGE_CHANNEL");
    public static final EventContextKey<LinkResult> LINK_RESULT = createFor("Nico-Yazawa:LINK_RESULT");


    @SuppressWarnings("unchecked")
    private static <T> EventContextKey<T> createFor(String id) {
        return DummyObjectProvider.createFor(EventContextKey.class, id);
    }

}

File: src/main/kotlin/com/nanabell/sponge/nico/link/event/LinkEventContextKeys.kt
package com.nanabell.sponge.nico.link.event

import com.nanabell.sponge.nico.extensions.DiscordUser
import com.nanabell.sponge.nico.extensions.MinecraftUser
import net.dv8tion.jda.api.entities.MessageChannel
import org.spongepowered.api.event.cause.EventContextKey
import org.spongepowered.api.util.generator.dummy.DummyObjectProvider

object LinkEventContextKeys {

    val DISCORD_USER = createFor<DiscordUser>("DISCORD_USER")
    val MINECRAFT_USER = createFor<MinecraftUser>("MINECRAFT_USER")
    val MESSAGE_CHANNEL = createFor<MessageChannel>("DISCORD_MESSAGE_CHANNEL")

    @Suppress("UNCHECKED_CAST")
    private fun <T> createFor(id: String): EventContextKey<T> {
        return DummyObjectProvider.createFor<EventContextKey<*>>(EventContextKey::class.java, id) as EventContextKey<T>
    }
}